src/app/CommandHandler.cpp

```diff
@@ -71,6 +71,8 @@ CHIP_ERROR CommandHandler::SendCommandResponse()
     MoveToState(CommandState::Sending);
 
 exit:
+    // Keep Shutdown() from double-closing our exchange.
+    mpExchangeCtx = nullptr;
```

Comment:
The exchange is auto closed on message receipt, how this exchange is closed after sending response?

Reply:
Right now we call `SendCommandResponse` before we return from `OnInvokeCommandRequest` and hence before the auto-closing happens.  If that changes, we would switch to calling `WillSendMessage` under `OnInvokeCommandRequest` and then close in `SendCommandResponse` (I plan to make this pretty automatic by closing under `SendMessage` in a followup PR, fwiw).

So if this line were not there, `Shutdown` would call `Abort` on the exchange and then the exchange itself would call `Close` on itself, messing up the reference counting.

Again, this will become much cleaner when we no longer need any explicit closing from outside the exchange, just clearly letting it know what our application state is.  This is the first step to getting there.

Reply:
Right now we use setting mpExchangeCtx = nullptr to prevent  double close and assume the exchange will be closed under hood. 

Since we do have a response to send, how about calling WillSendMessage in OnInvokeCommandRequest before SendCommandResponse and leave to Shutdown to close the exchange?

Reply:
That wouldn't really change anything, because the sequence of events would then presumably be:

1. `ExchangeContext::HandleMessage` is entered.
2. `OnInvokeCommandRequest` is called.
3. `WillSendMessage` is called.
4. `SendCommandResponse` is called and clears the "will send message" flag: the message that was promised has now been sent.
5. The `Shutdown` call in `SendCommandResponse` aborts the exchange.
6. The auto-close code is reached.

That is, `WillSendMessage` only makes sense if you plan to go async with your message sending.

That said, the fact that the auto-close code now checks whether the exchange is already closed (something I needed to add to deal with the way we hand acks to already-closed exchanges) means this pattern, or indeed not calling `WillSendMessage` and not nulling out `mpExchangeCtx` would be ok.  It feels to me that it would be a little confusing to have two different callsites that are both "responsible for closing the exchange", one of which just happens to not do anything if the other one has happened.  But I'd be OK with it as a temporary state of affairs.  Again, the goal is to move to a world where there is no manual closing of exchanges at all, just calls about application-level desires that then lead the exchange to be closed once it's no longer needed.

Reply:
Yes, the result will be the same. The tricky part now is the auto-close could be triggered in any case, the application needs to clearly figure out if the auto-close will be reached later and take action to prevent double close if needed. Such as by setting exchange to nullptr.

Do you think if we can make life easier if we enforce the following rules?
1. The exchange will always be auto-closed on message receipt if there is no more message needed to be sent.
2. If there is more message to be sent, application always set WillSendMessage flag, and manually close the exchange after extra message is sent.

Reply:
My long-term plan (of which this PR is the first step) is to have something like the following rules:

1. An exchange can be in one of three states: waiting for send, waiting for receive, or closed.
2. Calling SendMessage on the exchange transitions to "waiting for receive" if the "wait for response" flag is set, or to closed otherwise.
3. Calling HandleMessage on the exchange can only happen in the "waiting for receive" state and can land it in any of the three states, corresponding to the delegate calling WillSendMessage, calling SendMessage with the "wait for response" flag, or to the delegate calling neither.  Some exceptions here for things that are there just for CRMP (i.e. HandleMessage calls for acks and duplicates), which we might perhaps want to factor out into a different API, not HandleMessage.
4. Timing out an exchange that is in "waiting for receive" state has the same general behavior as HandleMessage.
4. Exchanges are not reference counted.  An exchange waiting for send is owned by the application.  An exchange waiting for receive or closed is owned by the CHIP stack.  
5. Ideally you can't have a reference to an exchange you don't own.  This part might require some handle classes that I have not thought completely through yet.

But that's long-term.  For purposes of this PR the two rules you describe are pretty much what we are aiming for, except that in addition to doing WillSendMessage and then manually closing you can just send a message that expects a response, which is one step closer to my goal rules.

Reply:
Those rules sound good to me, I am ok we only handle the auto-close on receipt with this PR as the first steps to achieve those goals.